tokio-util/src/task/spawn_pinned.rs

```diff
@@ -0,0 +1,248 @@
+    where
+        Fut::Output: Send + 'static,
+    {
+        let worker = self.find_and_incr_least_burdened_worker();
+
+        // Send the future to the worker
+        let (sender, receiver) = channel();
+        let task_count = Arc::clone(&worker.task_count);
+        let request = FutureRequest {
+            spawn: Box::new(move || {
+                let join_handle = spawn_local(async move {
+                    let result = create_task().await;
+
+                    // Update the task count once the future is finished
+                    task_count.fetch_sub(1, Ordering::SeqCst);
+
+                    result
+                });
+
+                sender.send(join_handle).ok();
+            }),
+        };
+        worker.spawner.send(request).unwrap();
+
+        receiver
+    }
+
+    /// Find the worker with the least number of tasks, increment its task
+    /// count, and return its handle. Make sure to actually spawn a task on
+    /// the worker so the task count is kept consistent with load.
+    fn find_and_incr_least_burdened_worker(&self) -> &LocalWorkerHandle {
+        loop {
+            let (worker, task_count) = self
+                .workers
+                .iter()
+                .map(|worker| (worker, worker.task_count.load(Ordering::SeqCst)))
+                .min_by_key(|&(_, count)| count)
+                .expect("There must be more than one worker");
+
+            // Make sure the task count hasn't changed when we choose this worker.
+            // Otherwise, restart the search.
+            if worker
+                .task_count
+                .compare_exchange(
+                    task_count,
+                    task_count + 1,
+                    Ordering::SeqCst,
+                    Ordering::Relaxed,
+                )
+                .is_ok()
+            {
+                return worker;
+            }
+        }
+    }
+}
+
+type PinnedFutureSpawner = Box<dyn FnOnce() + Send + 'static>;
+
+struct FutureRequest {
+    spawn: PinnedFutureSpawner,
+}
+
+// Needed for the unwrap in LocalPool::spawn_pinned if sending fails
+impl Debug for FutureRequest {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        f.write_str("FutureRequest")
+    }
+}
+
+struct LocalWorkerHandle {
+    spawner: UnboundedSender<FutureRequest>,
+    task_count: Arc<AtomicUsize>,
+}
+
+impl LocalWorkerHandle {
+    /// Create a new worker for executing pinned tasks
+    fn new_worker() -> LocalWorkerHandle {
+        let (sender, receiver) = unbounded_channel();
+        std::thread::spawn(|| Self::run(receiver));
+
+        LocalWorkerHandle {
+            spawner: sender,
+            task_count: Arc::new(AtomicUsize::new(0)),
+        }
+    }
+
+    fn run(mut task_receiver: UnboundedReceiver<FutureRequest>) {
+        let runtime = Builder::new_current_thread()
+            .enable_all()
+            .build()
+            .expect("Failed to start a pinned worker thread runtime");
```

Comment:
One option here would be to build it before spawning so that a panic would happen in the user's thread.

Reply:
Ok, sounds good. I was worried that creating it on one thread and then moving it to another would mess up some thread-local state. I'll try applying that refactor.